File: rust/src/diagnostic.rs
use std::sync::Arc;

use oxc::diagnostics::{LabeledSpan, NamedSource, OxcDiagnostic};
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct Diagnostic {
  pub severity: &'static str,
  pub message: String,
  pub labels: Vec<Label>,
  #[serde(skip_serializing_if = "Option::is_none")]
  pub help: Option<String>,
  #[serde(skip_serializing_if = "Option::is_none")]
  pub codeframe: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Label {
  #[serde(skip_serializing_if = "Option::is_none")]
  pub message: Option<String>,
  pub start: u32,
  pub end: u32,
}

impl Diagnostic {
  pub fn from_diagnostics(
    filename: &str,
    source_text: &str,
    diagnostics: impl IntoIterator<Item = OxcDiagnostic>,
  ) -> Vec<Self> {
    let diagnostics = diagnostics.into_iter().collect::<Vec<_>>();

    if diagnostics.is_empty() {
      return Vec::new();
    }

    let source = Arc::new(NamedSource::new(filename, source_text.to_string()));

    diagnostics
      .into_iter()
      .map(|diagnostic| Self::from_diagnostic(&source, diagnostic))
      .collect()
  }

  fn from_diagnostic(source: &Arc<NamedSource<String>>, diagnostic: OxcDiagnostic) -> Self {
    let severity = match diagnostic.severity {
      oxc::diagnostics::Severity::Error => "error",
      oxc::diagnostics::Severity::Warning => "warning",
      oxc::diagnostics::Severity::Advice => "advice",
    };

    let labels = diagnostic.labels.iter().map(Label::from).collect::<Vec<_>>();
    let message = diagnostic.message.to_string();
    let help = diagnostic.help.as_ref().map(ToString::to_string);
    let codeframe = Some(diagnostic.render_with_source_code(Arc::clone(source)));

    Self {
      severity,
      message,
      labels,
      help,
      codeframe,
    }
  }
}

impl From<&LabeledSpan> for Label {
  fn from(label: &LabeledSpan) -> Self {
    Self {
      message: label.label().map(ToString::to_string),
      start: label.offset(),
      end: label.offset() + label.len(),
    }
  }
}

File: lib/oxc/diagnostic.rb
# frozen_string_literal: true

module Oxc
  class Diagnostic
    ERROR = "error" #: String
    WARNING = "warning" #: String

    attr_reader :severity #: String
    attr_reader :message #: String
    attr_reader :labels #: Array[Oxc::Label]
    attr_reader :help #: String?
    attr_reader :codeframe #: String?

    #: (Hash[String, untyped]) -> Oxc::Diagnostic
    def self.from_hash(parsed)
      new(
        severity: parsed.fetch("severity"),
        message: parsed.fetch("message"),
        labels: parsed.fetch("labels").map { |label| Label.from_hash(label) },
        help: parsed["help"],
        codeframe: parsed["codeframe"]
      )
    end

    #: (severity: String, message: String, labels: Array[Oxc::Label], ?help: String?, ?codeframe: String?) -> void
    def initialize(severity:, message:, labels:, help: nil, codeframe: nil)
      @severity = severity
      @message = message
      @labels = labels.freeze
      @help = help
      @codeframe = codeframe

      freeze
    end

    #: () -> bool
    def error?
      severity == ERROR
    end

    #: () -> bool
    def warning?
      severity == WARNING
    end

    #: () -> String
    def to_s
      message
    end

    #: () -> String
    def inspect
      "#<#{self.class.name} #{severity} #{message.inspect}>"
    end
  end

  class Label
    attr_reader :message #: String?
    attr_reader :start #: Integer
    attr_reader :finish #: Integer

    #: (Hash[String, untyped]) -> Oxc::Label
    def self.from_hash(parsed)
      new(message: parsed["message"], start: parsed.fetch("start"), finish: parsed.fetch("end"))
    end

    #: (start: Integer, finish: Integer, ?message: String?) -> void
    def initialize(start:, finish:, message: nil)
      @start = start
      @finish = finish
      @message = message

      freeze
    end

    #: (String) -> String?
    def slice(source)
      source.byteslice(start, finish - start)
    end

    #: () -> String
    def inspect
      "#<#{self.class.name} range=[#{start}, #{finish}]#{" #{message.inspect}" if message}>"
    end
  end
end
